metagraph/src/graph/alignment/aligner_methods.hpp

```diff
@@ -13,203 +13,187 @@
-    typedef typename Seeder<NodeType>::Seed Seed;
-
-    SuffixSeeder(const DeBruijnGraph &graph, const DBGAlignerConfig &config);
-
-    void initialize(std::string_view query, bool orientation);
-
-    void call_seeds(std::function<void(Seed&&)> callback) const;
-
-    const DeBruijnGraph& get_graph() const { return base_seeder_->get_graph(); }
-    virtual const std::string_view get_query() const { return base_seeder_->get_query(); }
-    const DBGAlignerConfig& get_config() const { return base_seeder_->get_config(); }
-    bool get_orientation() const { return base_seeder_->get_orientation(); }
+    typedef Alignment<NodeType> Seed;
 
-  private:
-    std::unique_ptr<ExactMapSeeder<NodeType>> base_seeder_;
-    std::vector<std::vector<NodeType>> alt_query_nodes;
+    virtual ~ISeeder() {}
 
-    std::vector<NodeType>& get_query_nodes() { return base_seeder_->get_query_nodes(); }
-    std::vector<uint8_t>& get_offsets() { return base_seeder_->get_offsets(); }
-    size_t get_num_matching_kmers() const { return base_seeder_->get_num_matching_kmers(); }
+    virtual void call_seeds(std::function<void(Seed&&)> callback) const = 0;
 };
 
 template <typename NodeType = typename DeBruijnGraph::node_index>
-class ExactMapSeeder : public Seeder<NodeType> {
-    friend SuffixSeeder<NodeType>;
-
+class ExactMapSeeder : public ISeeder<NodeType> {
   public:
-    typedef typename Seeder<NodeType>::Seed Seed;
-    typedef DBGAlignerConfig::score_t score_t;
+    typedef typename ISeeder<NodeType>::Seed Seed;
 
-    ExactMapSeeder(const DeBruijnGraph &graph, const DBGAlignerConfig &config)
-          : graph_(graph), config_(config) { assert(config_.check_config_scores()); }
+    ExactMapSeeder(const DeBruijnGraph &graph,
+                   std::string_view query,
+                   bool orientation,
+                   std::vector<NodeType>&& nodes,
+                   const DBGAlignerConfig &config);
 
     virtual ~ExactMapSeeder() {}
 
-    virtual const DeBruijnGraph& get_graph() const override final { return graph_; }
-    virtual const DBGAlignerConfig& get_config() const override final { return config_; }
+  protected:
+    const std::string_view get_query() const { return query_; }
+    bool get_orientation() const { return orientation_; }
 
-    const std::string_view get_query() const override { return query_; }
     const std::vector<NodeType>& get_query_nodes() const { return query_nodes_; }
-    const std::vector<uint8_t>& get_offsets() const { return offsets_; }
-    const std::vector<score_t>& get_partial_sums() const { return partial_sum_; }
-    bool get_orientation() const override { return orientation_; }
+    const std::vector<DBGAlignerConfig::score_t>& get_partial_sums() const {
+        return partial_sum_;
+    }
 
-    virtual void initialize(std::string_view query, bool orientation) override;
+    const DeBruijnGraph& get_graph() const { return graph_; }
+    const DBGAlignerConfig& get_config() const { return config_; }
```

Comment:
Simply move `graph_` and `config_`to protected so you can access them directly and remove these methods.

Reply:
done